File: 김순석/프로그래머스/pro_키패드누르기.py
dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]


def solution(numbers, hand):
    answer = ''
    L, R = (3, 0), (3, 2)

    for i in range(len(numbers)):
        if numbers[i] == 1 or numbers[i] == 4 or numbers[i] == 7:
            L = (numbers[i] // 3, 0)
            answer += 'L'

        if numbers[i] == 3 or numbers[i] == 6 or numbers[i] == 9:
            R = ((numbers[i] // 3) - 1, 2)
            answer += 'R'

        if numbers[i] == 2 or numbers[i] == 5 or numbers[i] == 8:
            target = (numbers[i] // 3, 1)
            ans, l, r = calc(L, R, target, hand)
            answer += ans
            L, R = l, r

        if numbers[i] == 0:
            target = (3, 1)
            ans, l, r = calc(L, R, target, hand)
            answer += ans
            L, R = l, r
    return answer


def calc(l, r, t, hand):
    xl, yl, xr, yr = l[0], l[1], r[0], r[1]
    tx, ty = t[0], t[1]

    l_dist = abs(tx - xl) + abs(ty - yl)
    r_dist = abs(tx - xr) + abs(ty - yr)

    if l_dist == r_dist:

        if hand == 'left':
            l = t
            return 'L', l, r
        else:
            r = t
            return 'R', l, r

    if l_dist > r_dist:
        r = t
        return 'R', l, r
    if l_dist < r_dist:
        l = t
        return 'L', l, r

    return

numbers = [7, 0, 8, 2, 8, 3, 1, 5, 7, 6, 2]
hand = "left"
print(solution(numbers, hand))
print()

File: AHYEON/12.programers_lv1_lv2/pro_키패드누르기.swift
import Foundation

func solution(_ numbers:[Int], _ hand:String) -> String {
    var left: (Int, Int) = (3, 0)
    var right: (Int, Int) = (3, 2)
    var answer: String = ""

    let pos: [(Int, Int)] = [
        (3, 1), (0, 0), (0, 1), (0, 2),
        (1, 0), (1, 1), (1, 2),
        (2, 0), (2, 1), (2, 2)
    ]

    numbers.forEach {
        if $0 == 1 || $0 == 4 || $0 == 7 {
            left = pos[$0]
            answer += "L"
        }
        else if $0 == 3 || $0 == 6 || $0 == 9 {
            right = pos[$0]
            answer += "R"
        }
        else {
            let leftDist: Int = abs(left.0 - pos[$0].0) + abs(left.1 - pos[$0].1)
            let rightDist: Int = abs(right.0 - pos[$0].0) + abs(right.1 - pos[$0].1)
            if leftDist == rightDist {
                if hand == "right" {
                    answer += "R"
                    right = pos[$0]
                }
                else {
                    answer += "L"
                    left = pos[$0]
                }
            } else if leftDist > rightDist {
                answer += "R"
                right = pos[$0]
            } else {
                answer += "L"
                left = pos[$0]
            }
        }

    }

    return answer
}